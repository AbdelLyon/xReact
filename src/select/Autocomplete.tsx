// import { useState, useCallback } from "react";
// import { useInView } from "react-intersection-observer";

// import { ErrorMessage } from "../shared/input/ErrorMessage";
// import { LoadingSpinner } from "../shared/input/LoadingSpinner";
// import { SelectTrigger as Trigger } from "../shared/select/SelectTrigger";

// import { cn } from "@/utils";
// import { Popover, PopoverContent } from "@/ui/popover";
// import { useInfiniteScroll } from "@/hooks/useInfiniteScroll";
// import { SearchInput } from "@/shared/select/SearchInput";
// import { Options } from "@/shared/select/Options";
// import { SelectProps } from "@/types/formTypes";
// import {
//   SelectGroup,
//   SelectValue,
//   SelectTrigger,
//   SelectContent,
//   SelectLabel,
//   SelectItem,
// } from "@/ui/select";

// const Select = ({
//   options,
//   label,
//   placeholder,
//   isSearchable,
//   width,
//   error,
//   isFetchingPage,
//   onFetchPage,
//   searchPlaceholder,
//   selectedKeys = [],
//   classNames = {},
//   value,
//   onSelect,
//   onSearch,
//   allOptionsLabel,
// }: SelectProps) => {
//   const { ref, inView } = useInView();
//   const [isOpen, setIsOpen] = useState(false);

//   useInfiniteScroll(inView, onFetchPage);

//   const handleOpenChange = useCallback((open: boolean) => {
//     setIsOpen(open);
//   }, []);

//   return (
//     <div data-testid="select">
//       <Popover open={isOpen} onOpenChange={handleOpenChange}>
//         {label && <label className="text-start">{label}</label>}
//         <Trigger
//           className={cn("w-full", classNames.trigger)}
//           error={error}
//           options={options}
//           placeholder={placeholder}
//           selectedKeys={selectedKeys}
//           value={value}
//           width={width}
//         />
//         <PopoverContent
//           className={cn("bg-white dark:bg-card", classNames.popover)}
//           style={{
//             width: width ?? "100%",
//           }}
//         >
//           {isSearchable && (
//             <SearchInput placeholder={searchPlaceholder} onSearch={onSearch} />
//           )}
//           <div className="relative">
//             <Options
//               allOptionsLabel={allOptionsLabel}
//               className={cn("bg-white dark:bg-card", classNames.popover)}
//               observerRef={ref}
//               options={options}
//               placeholder={placeholder}
//               selectedKeys={selectedKeys}
//               setIsOpen={setIsOpen}
//               value={value}
//               onSelect={onSelect}
//             />
//             {isFetchingPage && <LoadingSpinner />}
//           </div>
//         </PopoverContent>
//         {error && <ErrorMessage error={error} />}
//       </Popover>
//     </div>
//   );
// };

// export {
//   Select,
//   SelectGroup,
//   SelectValue,
//   SelectTrigger,
//   SelectContent,
//   SelectLabel,
//   SelectItem,
// };

import {
  Autocomplete as AutocompleteComponent,
  AutocompleteItem,
  AutocompleteProps,
  SlotsToClasses,
} from "@nextui-org/react";
import { Key } from "@react-types/shared";

import { cn } from "@/utils";

// Type générique avec contraintes plus strictes
type GenericItem = {
  key: Key;
  label: string;
  [key: string]: any;
};

type GenericAutocompleteProps<T extends GenericItem> = {
  items: T[];
  labelKey?: keyof T;
  valueKey?: keyof T;
  descriptionKey?: keyof T;
  autocompleteProps?: Partial<AutocompleteProps<T>>;
  itemRenderer?: (item: T) => React.ReactElement;
  onSelectionChange?: (key: Key | null) => void;
  defaultSelectedKey?: Key;
  placeholder?: string;
  label?: string;
  description?: string;
  classNames?: SlotsToClasses<
    | "base"
    | "listbox"
    | "listboxWrapper"
    | "popoverContent"
    | "endContentWrapper"
    | "clearButton"
    | "selectorButton"
  >;
};

export function Autocomplete<T extends GenericItem>({
  items,
  labelKey = "label" as keyof T,
  valueKey = "key" as keyof T,
  descriptionKey = "description" as keyof T,
  autocompleteProps,
  itemRenderer,
  onSelectionChange,
  defaultSelectedKey,
  placeholder = "Search...",
  label,
  description,
  classNames,
}: GenericAutocompleteProps<T>) {
  const defaultItemRenderer = (item: T) => (
    <AutocompleteItem
      key={item[valueKey]}
      description={(item[descriptionKey] as string) || undefined}
    >
      {item[labelKey] as string}
    </AutocompleteItem>
  );

  return (
    <AutocompleteComponent
      className="max-w-xs"
      classNames={{
        popoverContent: cn("bg-background", classNames?.popoverContent),

        ...classNames,
      }}
      defaultItems={items}
      defaultSelectedKey={defaultSelectedKey}
      description={description}
      label={label}
      placeholder={placeholder}
      onSelectionChange={onSelectionChange}
      {...autocompleteProps}
    >
      {(item: T) =>
        itemRenderer ? itemRenderer(item) : defaultItemRenderer(item)
      }
    </AutocompleteComponent>
  );
}
