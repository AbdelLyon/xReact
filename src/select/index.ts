import "@/index.css";
export { Autocomplete } from "@/select/Autocomplete";
