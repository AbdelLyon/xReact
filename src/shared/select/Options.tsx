import { useCallback } from "react";
import { Check } from "lucide-react";

import { cn } from "@/utils";
import {
  Command,
  CommandEmpty,
  CommandGroup,
  CommandItem,
  CommandList,
} from "@/ui/command";
import { SelectProps } from "@/types/formTypes";
const Option = ({
  options,
  value,
  observerRef,
  handleSelect,
  selectedKeys,
}: Partial<SelectProps> & {
  handleSelect: (currentValue: string) => void;
}) => {
  return options?.map((item, index) => {
    const itemId = String(item?.id);
    const isSelected = value === item?.id;

    return (
      <CommandItem
        key={itemId}
        ref={options.length === index + 1 ? observerRef : null}
        className="flex justify-between py-2"
        value={itemId}
        onSelect={() => handleSelect(itemId)}
      >
        <p className="truncate">
          {selectedKeys?.map((key) => item[key] ?? "").join(" ")}
        </p>
        <Check
          className={cn(
            "mr-2 h-4 w-4",
            isSelected ? "text-green-600/80 opacity-100" : "opacity-0",
          )}
        />
      </CommandItem>
    );
  });
};

export const Options = ({
  options = [],
  onSelect,
  value,
  setIsOpen,
  selectedKeys = [],
  observerRef,
  className,
  allOptionsLabel,
}: SelectProps) => {
  const handleSelect = useCallback(
    (currentValue: string) => {
      const optionSelected = options.find(
        ({ id }) => String(id) === currentValue,
      );

      if (optionSelected) onSelect(optionSelected);

      if (setIsOpen) setIsOpen(false);
    },
    [onSelect, options, value, setIsOpen],
  );

  return (
    <Command className={cn("mb-1 bg-background dark:bg-card", className)}>
      <CommandList>
        {options.length === 0 && (
          <CommandEmpty className="hover:bg-light mx-1 mt-1 h-9 p-3 text-sm dark:hover:bg-background">
            Aucune option
          </CommandEmpty>
        )}

        {allOptionsLabel && (
          <CommandItem
            className="hover:bg-light mx-1 mt-1 h-9 p-3 text-sm dark:hover:bg-background"
            value=""
            onSelect={() => handleSelect("")}
          >
            {allOptionsLabel}
          </CommandItem>
        )}

        <CommandGroup>
          <Option
            handleSelect={handleSelect}
            observerRef={observerRef}
            options={options}
            selectedKeys={selectedKeys}
            value={value}
          />
        </CommandGroup>
      </CommandList>
    </Command>
  );
};
