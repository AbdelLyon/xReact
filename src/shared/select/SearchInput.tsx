import { useCallback } from "react";

import { Input } from "@/input";

type Props = {
  placeholder?: string;
  onSearch?: (value: string) => void;
};
export const SearchInput = ({ placeholder, onSearch }: Props) => {
  const handleSearchChange = useCallback(
    (e: React.ChangeEvent<HTMLInputElement>) => onSearch?.(e.target.value),
    [onSearch],
  );

  return (
    <div className="m-2">
      <Input
        name="search"
        placeholder={placeholder}
        type="search"
        onChange={handleSearchChange}
      />
    </div>
  );
};
