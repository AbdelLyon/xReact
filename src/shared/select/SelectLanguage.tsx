import { Language, StaticImageData } from "@/types/layoutTypes";
import {
  Select,
  SelectContent,
  SelectGroup,
  SelectItem,
  SelectTrigger,
  SelectValue,
} from "@/ui/select";
import { cn } from "@/utils";

export const SelectLanguage = ({
  value,
  languages,
  onLanguageChange,
}: {
  value: string;
  languages?: Language[];
  onLanguageChange?: (language: Language) => void;
}) => {
  return (
    <Select
      value={value ?? "fr"}
      onValueChange={(value) => {
        const selectedLanguage = languages?.find((l) => l.value === value);

        if (selectedLanguage) onLanguageChange?.(selectedLanguage);
      }}
    >
      <SelectTrigger
        className={cn(
          "focus:ring-0 shadow-none bg-transparent dark:text-white border-none w-max",
          { "text-muted-foreground": value === "" },
        )}
      >
        <SelectValue placeholder={value} />
      </SelectTrigger>
      <SelectContent>
        <SelectGroup>
          {languages?.map(
            (country: {
              flag?: StaticImageData;
              label?: string;
              value: string;
            }) => (
              <SelectItem key={country.value} value={country.value}>
                <img
                  alt={country.label ?? country.value}
                  className="h-[14px] w-[18px] rounded-sm"
                  src={country?.flag?.src}
                />
              </SelectItem>
            ),
          )}
        </SelectGroup>
      </SelectContent>
    </Select>
  );
};
