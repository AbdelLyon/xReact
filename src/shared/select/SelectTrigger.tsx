import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faChevronDown } from "@fortawesome/free-solid-svg-icons";

import { cn } from "@/utils";
import { PopoverTrigger } from "@/ui/popover";
import { SelectProps } from "@/types/formTypes";

export const SelectTrigger = ({
  className,
  value,
  selectedKeys = [],
  placeholder,
  width,
  options = [],
  error = "",
}: Partial<SelectProps>) => {
  const valueSelected = options?.find((item) => item?.id === value) ?? {};

  return (
    <PopoverTrigger asChild>
      <div
        className={cn(
          "flex h-11 cursor-pointer items-center justify-between rounded-md border border-border bg-white dark:bg-background px-3 relative",
          {
            "border-error": error !== "",
          },
          className,
        )}
        style={{
          width: width ?? "100%",
        }}
      >
        <p
          className={cn("truncate text-[13px]", {
            "opacity-50": !value,
          })}
        >
          {value
            ? selectedKeys?.map((s) => `${valueSelected?.[s]} `)
            : placeholder}
        </p>

        <FontAwesomeIcon
          className={cn("absolute right-3 shrink-0 opacity-50", {
            "text-error opacity-100": error !== "",
          })}
          icon={faChevronDown}
        />
      </div>
    </PopoverTrigger>
  );
};
