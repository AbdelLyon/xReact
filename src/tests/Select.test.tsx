import { render } from "@testing-library/react";
import { describe, it, expect } from "vitest";

import { Select } from "@/select";

describe("Select", () => {
  it("should match the snapshot with default props", () => {
    const options = [
      { label: "Option 1", value: "1" },
      { label: "Option 2", value: "2" },
    ];

    const { asFragment } = render(
      <Select
        isFetchingPage={false}
        label="Select an option"
        options={options}
        placeholder="Select..."
        selectedKeys={[]}
        onFetchPage={() => {}}
        onSearch={() => {}}
        onSelect={() => {}}
      />,
    );

    expect(asFragment()).toMatchSnapshot();
  });
  it("should match the snapshot with search enabled and loading", () => {
    const options: Array<{ label: string; value: string }> = [
      { label: "Option 1", value: "1" },
      { label: "Option 2", value: "2" },
    ];

    const { asFragment } = render(
      <Select
        isSearchable
        isFetchingPage={true}
        label="Searchable Select"
        options={options}
        placeholder="Search..."
        searchPlaceholder="Search options..."
        selectedKeys={[]}
        onFetchPage={() => {}}
        onSearch={() => {}}
        onSelect={() => {}}
      />,
    );

    expect(asFragment()).toMatchSnapshot();
  });
  it("should match the snapshot with an error", () => {
    const options = [
      { label: "Option 1", value: "1" },
      { label: "Option 2", value: "2" },
    ];

    const { asFragment } = render(
      <Select
        error="Something went wrong"
        isFetchingPage={false}
        label="Select with error"
        options={options}
        placeholder="Select..."
        selectedKeys={[]}
        onFetchPage={() => {}}
        onSearch={() => {}}
        onSelect={() => {}}
      />,
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
