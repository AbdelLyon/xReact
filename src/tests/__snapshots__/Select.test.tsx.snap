// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Select > should match the snapshot with an error 1`] = `
<DocumentFragment>
  <div
    data-testid="select"
  >
    <label
      class="text-start"
    >
      Select with error
    </label>
    <div
      aria-controls="radix-:r2:"
      aria-expanded="false"
      aria-haspopup="dialog"
      class="flex h-11 cursor-pointer items-center justify-between rounded-md border bg-white dark:bg-background px-3 relative border-error w-full"
      data-state="closed"
      style="width: 100%;"
      type="button"
    >
      <p
        class="truncate text-[13px] opacity-50"
      >
        Select...
      </p>
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-chevron-down absolute right-3 shrink-0 text-error opacity-100"
        data-icon="chevron-down"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M233.4 406.6c12.5 12.5 32.8 12.5 45.3 0l192-192c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L256 338.7 86.6 169.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l192 192z"
          fill="currentColor"
        />
      </svg>
    </div>
    <p
      class="text-error mt-1 text-start"
    >
      Something went wrong
    </p>
  </div>
</DocumentFragment>
`;

exports[`Select > should match the snapshot with default props 1`] = `
<DocumentFragment>
  <div
    data-testid="select"
  >
    <label
      class="text-start"
    >
      Select an option
    </label>
    <div
      aria-controls="radix-:r0:"
      aria-expanded="false"
      aria-haspopup="dialog"
      class="flex h-11 cursor-pointer items-center justify-between rounded-md border border-border bg-white dark:bg-background px-3 relative w-full"
      data-state="closed"
      style="width: 100%;"
      type="button"
    >
      <p
        class="truncate text-[13px] opacity-50"
      >
        Select...
      </p>
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-chevron-down absolute right-3 shrink-0 opacity-50"
        data-icon="chevron-down"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M233.4 406.6c12.5 12.5 32.8 12.5 45.3 0l192-192c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L256 338.7 86.6 169.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l192 192z"
          fill="currentColor"
        />
      </svg>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Select > should match the snapshot with search enabled and loading 1`] = `
<DocumentFragment>
  <div
    data-testid="select"
  >
    <label
      class="text-start"
    >
      Searchable Select
    </label>
    <div
      aria-controls="radix-:r1:"
      aria-expanded="false"
      aria-haspopup="dialog"
      class="flex h-11 cursor-pointer items-center justify-between rounded-md border border-border bg-white dark:bg-background px-3 relative w-full"
      data-state="closed"
      style="width: 100%;"
      type="button"
    >
      <p
        class="truncate text-[13px] opacity-50"
      >
        Search...
      </p>
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-chevron-down absolute right-3 shrink-0 opacity-50"
        data-icon="chevron-down"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M233.4 406.6c12.5 12.5 32.8 12.5 45.3 0l192-192c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L256 338.7 86.6 169.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l192 192z"
          fill="currentColor"
        />
      </svg>
    </div>
  </div>
</DocumentFragment>
`;
